src/contexts/EditorDOMEventRegistryContext.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  DependencyList,
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+} from "react";
+
+import { useForceUpdate } from "../hooks/useForceUpdate";
+import {
+  DOMEventMap,
+  EventHandler,
+  createReactEventPlugin,
+} from "../plugins/createReactEventPlugin";
+
+/**
+ * Produces a plugin that can be used with ProseMirror to handle DOM
+ * events at the EditorView.dom element.
+ *
+ * - `reactEventsPlugin` is a ProseMirror plugin for handling DOM events
+ * at the EditorView.dom element. It should be passed to `useEditorView`,
+ * along with any other plugins.
+ *
+ * - 'registry' should be passed as the value to
+ * `EditorDOMEventRegistryContext`.
+ */
+export function useReactEventPlugin() {
+  const registry = useRef(
+    new Map<keyof DOMEventMap, Set<EventHandler>>()
+  ).current;
+
+  const forceUpdate = useForceUpdate();
+
+  const register = useCallback(
+    (eventType: keyof DOMEventMap, handler: EventHandler) => {
+      const handlers = registry.get(eventType) ?? new Set<EventHandler>();
+      handlers.add(handler);
+      if (!registry.has(eventType)) {
+        // We only need to force an update when the set of event types
+        // that we care about changes
+        registry.set(eventType, handlers);
+        forceUpdate();
+      }
+    },
+    [registry, forceUpdate]
+  );
+
+  const unregister = useCallback(
+    (eventType: keyof DOMEventMap, handler: EventHandler) => {
+      const handlers = registry.get(eventType);
+      if (!handlers) return;
+      handlers.delete(handler);
+    },
+    [registry]
+  );
+
+  const reactEventPlugin = useMemo(
+    () => createReactEventPlugin(registry),
+    // We force an update whenever the set of event types that
+    // we care about changes. This plugin only needs to be
+    // recreated when that set (the set of registry keys)
+    // changes.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [registry.keys()]
```

Comment:
This returns a new iterable every time. We could spread it, because the order should be stable, but maybe we can just do something more obvious. For example, instead of `forceUpdate`, we could the registry in state and set it to a new object when we add/remove keys (but mutate its values in place when we don't).

Reply:
.. whoops. Yes, lemme try that!